nautobot/ipam/tests/test_models.py

```diff
@@ -59,6 +59,64 @@ def setUpTestData(cls):
             status=vmint_status,
         )
 
+    def test_removing_ip_addresses_containing_host_device_primary_ip_not_allowed(self):
+        """
+        Validate we cannot remove an IPAddress from an Interface that is the host Device's primary ip.
+        """
+        dev_ip_addr = IPAddress.objects.create(address="192.0.2.1/24")
+        self.test_int1.add_ip_addresses(dev_ip_addr)
+        ip_to_interface = IPAddressToInterface.objects.get(interface=self.test_int1, ip_address=dev_ip_addr)
+        self.assertIsNotNone(ip_to_interface)
+        self.test_device.primary_ip4 = dev_ip_addr
+        self.test_device.save()
+        with self.assertRaises(ValidationError) as cm:
+            ip_to_interface.delete()
+        self.assertIn(
+            f"Cannot remove IP address {self.test_device.primary_ip4} from interface {self.test_int1} on Device {self.test_device.name} because it is marked as its primary IPv{self.test_device.primary_ip4.family} address",
+            str(cm.exception),
+        )
+
+    def test_removing_ip_addresses_containing_host_virtual_machine_primary_ip_not_allowed(self):
+        """
+        Validate we cannot remove an IPAddress from an Interface that is the host Virtual Machine's primary ip.
+        """
+        vm_ip_addr = IPAddress.objects.create(address="192.0.2.1/24")
+        self.test_vmint1.add_ip_addresses(vm_ip_addr)
+        ip_to_vminterface = IPAddressToInterface.objects.get(vm_interface=self.test_vmint1, ip_address=vm_ip_addr)
+        self.assertIsNotNone(ip_to_vminterface)
+        self.test_vm.primary_ip4 = vm_ip_addr
+        self.test_vm.save()
+        with self.assertRaises(ValidationError) as cm:
+            ip_to_vminterface.delete()
```

Comment:
Is there a `self.test_vmint1.remove_ip_addresses` API we should be using instead?